pkg/cidata/cidata.TEMPLATE.d/boot/04-persistent-data-volume.sh

```diff
@@ -61,21 +61,37 @@ if [ "$(awk '$2 == "/" {print $3}' /proc/mounts)" == "tmpfs" ]; then
 				PART=$(lsblk --list /dev/"${DISK}" --noheadings --output name,type | awk '$2 == "part" {print $1}')
 				mkfs.ext4 -L data-volume /dev/"${PART}"
 				mount -t ext4 /dev/disk/by-label/data-volume /mnt/data
-				# Unmount all mount points under /tmp so we can move it to the data volume:
-				# "mount1 on /tmp/lima type 9p (rw,dirsync,relatime,mmap,access=client,trans=virtio)"
-				for MP in $(mount | awk '$3 ~ /^\/tmp\// {print $3}'); do
-					umount "${MP}"
-				done
 				# setup apk package cache
 				mkdir -p /mnt/data/apk/cache
 				mkdir -p /etc/apk
 				ln -s /mnt/data/apk/cache /etc/apk/cache
+				# Prepare mnt.sh (used for restoring mounts later)
+				echo "#!/bin/sh" >/mnt/data/mnt.sh
+				echo "set -eux" >>/mnt/data/mnt.sh
 				# Move all persisted directories to the data volume
 				for DIR in ${DATADIRS}; do
 					DEST="/mnt/data$(dirname "${DIR}")"
 					mkdir -p "${DIR}" "${DEST}"
+					while IFS= read -r LINE; do
+						[ -z "$LINE" ] && continue
+						MNTDEV="$(echo "${LINE}" | awk '{print $1}')"
+						# unmangle " \t\n\\#"
+						# https://github.com/torvalds/linux/blob/v6.6/fs/proc_namespace.c#L89
+						MNTPNT="$(echo "${LINE}" | awk '{print $2}' | sed -e 's/\\040/ /g; s/\\011/\t/g; s/\\012/\n/g; s/\\134/\\/g; s/\\043/#/g')"
+						# This comparison needs [[, not [
+						[[ ${MNTPNT} != "${DIR}*" ]] && continue
```

Comment:
I think the `*` must be outside the quotes to be treated as a wildcard:

```suggestion
						[[ ${MNTPNT} != "${DIR}"* ]] && continue
```

But I don't think this works in `ash` (our `bash` is really just a wrapper for busybox `sh`, which is `ash` with some `bash` compatibility compiled in, but not everything works).

I would try something like this:

```shell
# Ignore if MNTPNT is neither DIR nor a parent directory of DIR.
# It is not a parent if MNTPNT doesn't start with DIR, or the first
# character after DIR isn't a slash.
WITHOUT_DIR=${MNTPNT#$DIR}
[ "$MNTPNT" != "$DIR" ] && [ "$MNTPNT" == "$WITHOUT_DIR" -o "${WITHOUT_DIR::1}" != "/" ] && continue
```

This also handles the case correctly when MNTPNT is `/tmpdir`, which would match `/tmp*`, but is not a subdirectory of it. Well, this is untested, so please double-check the condition!

Reply:
Thanks, tested and it seems to work.